upgrades/steps_28.go

```diff
@@ -72,3 +88,103 @@ func prepopulateRebootHandledFlagsForDeployedUnits(ctx Context) error {
 	}
 	return nil
 }
+
+func moveUniterStateToController(ctx Context) error {
+	// Lookup the names of all unit agents installed on this machine.
+	agentConf := ctx.AgentConfig()
+	_, unitNames, _, err := service.FindAgents(agentConf.DataDir())
+	if err != nil {
+		return errors.Annotate(err, "looking up unit agents")
+	}
+	if len(unitNames) == 0 {
+		// No units, nothing to do.
+		return nil
+	}
+
+	var fileNames []string
+	if ctx.AgentConfig().Value(agent.ProviderType) == k8sprovider.CAASProviderType {
+		if fileNames, err = caasOperatorLocalState(ctx, unitNames); err != nil {
+			return err
+		}
+	} else {
+		if fileNames, err = iaasUniterState(ctx, unitNames); err != nil {
+			return err
+		}
+	}
+
+	// Saving uniter state in the controller succeeded, now clean up
+	// the now unused state files.
+	for _, name := range fileNames {
+		if err = os.RemoveAll(name); err != nil {
+			// Log the error, but no need to fail the upgrade.  Juju
+			// will not use the file any longer.
+			logger.Errorf("failed to remove %q: %s", name, err)
+		}
+	}
+
+	return nil
+}
+
+// getUpgradeStepsClient is to facilitate mocking the unit tests.
+var getUpgradeStepsClient = func(caller base.APICaller) UpgradeStepsClient {
+	return upgradesteps.NewClient(caller)
+}
+
+//go:generate mockgen -package mocks -destination mocks/upgradestepsclient_mock.go github.com/juju/juju/upgrades UpgradeStepsClient
+type UpgradeStepsClient interface {
+	WriteUniterState(uniterStates map[names.Tag]string) error
+}
+
+func iaasUniterState(ctx Context, unitNames []string) ([]string, error) {
+	// Read the uniter state for each unit on this machine.
+	// Leave in yaml format as a string to push to the controller.
+	fileNames := []string{}
+	uniterStates := make(map[names.Tag]string, len(unitNames))
+	dataDir := ctx.AgentConfig().DataDir()
+	for _, unitName := range unitNames {
+		tag, err := names.ParseUnitTag(unitName)
+		if err != nil {
+			return nil, errors.Annotatef(err, "unable to parse unit agent tag %q", unitName)
+		}
+
+		// e.g. /var/lib/juju/agents/unit-ubuntu-0/state/uniter
+		uniterStateFile := filepath.Join(agent.BaseDir(dataDir), unitName, "state", "uniter")
+
+		// Read into a State as we can validate the data.
+		var st operation.State
+		if err = utils.ReadYaml(uniterStateFile, &st); err != nil {
+			// No file available, move on.
+			if os.IsNotExist(err) {
+				continue
+			}
+			return nil, errors.Annotatef(err, "failed to read uniter state from %q", uniterStateFile)
+		}
+		fileNames = append(fileNames, uniterStateFile)
+		if err = st.Validate(); err != nil {
+			return nil, errors.Annotatef(err, "failed to validate uniter state from %q", uniterStateFile)
+		}
+
+		data, err := yaml.Marshal(st)
```

Comment:
Sucks that we cannot pass a buffered reader to utils.ReadYaml so we could read the serialized data once instead of unmarshaling and marshaling...